src/opencl_androidbackup_fmt_plug.c

```diff
@@ -0,0 +1,379 @@
+/*
+ * JtR OpenCL format to crack Android Backup files.
+ *
+ * This software is Copyright (c) 2018 Dhiru Kholia, Copyright (c) 2018 magnum
+ * and it is hereby released to the general public under the following terms:
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * The OpenCL boilerplate code is borrowed from other OpenCL formats.
+ */
+
+#ifdef HAVE_OPENCL
+
+#include "arch.h"
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_opencl_ab;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_opencl_ab);
+#else
+
+#include <string.h>
+#include <stdint.h>
+
+#include "formats.h"
+#include "common.h"
+#include "androidbackup_common.h"
+#include "options.h"
+#include "jumbo.h"
+#include "opencl_common.h"
+#include "misc.h"
+#define MAX_OUTLEN 32
+#include "opencl_pbkdf2_hmac_sha1.h"
+
+#define FORMAT_LABEL            "AndroidBackup-opencl"
+#define ALGORITHM_NAME          "PBKDF2-SHA1 AES OpenCL"
+#define BENCHMARK_COMMENT       ""
+#define BENCHMARK_LENGTH        -1
+#define MIN_KEYS_PER_CRYPT      1
+#define MAX_KEYS_PER_CRYPT      1
+
+#define BINARY_SIZE             0
+#define BINARY_ALIGN            MEM_ALIGN_WORD
+#define PLAINTEXT_LENGTH        64
+#define SALT_SIZE               sizeof(*cur_salt)
+#define SALT_ALIGN              MEM_ALIGN_WORD
+
+typedef struct {
+	unsigned int cracked;
+} ab_out;
+
+typedef struct {
+	pbkdf2_salt pbkdf2;
+	uint32_t masteykey_blob_length;
+	unsigned char iv[IVLEN];
+	unsigned char masteykey_blob[MAX_MASTERKEYBLOB_LEN];
+} ab_salt;
+
+/* This handles all widths */
+#define GETPOS(i, index)        (((index) % ocl_v_width) * 4 + ((i) & ~3U) * ocl_v_width + (((i) & 3) ^ 3) + ((index) / ocl_v_width) * 64 * ocl_v_width)
+
+static struct custom_salt *cur_salt;
+static size_t key_buf_size;
+static unsigned int *inbuffer;
+static ab_out *output;
+static ab_salt currentsalt;
+static cl_mem mem_in, mem_dk, mem_salt, mem_state, mem_out;
+static size_t key_buf_size;
+static int new_keys;
+static struct fmt_main *self;
+
+static cl_kernel pbkdf2_init, pbkdf2_loop, pbkdf2_final, ab_final;
+
+/*
+ * HASH_LOOPS is ideally made by factors of (iteration count - 1) and should
+ * be chosen for a kernel duration of not more than 200 ms
+ */
+#define HASH_LOOPS              1024
+#define ITERATIONS              4098 /* Just for auto tune */
```

Comment:
"Just" for autotune? An incorrect figure here will result in suboptimal autotune. Since all of your test vectors seem to use 10000 I'm guessing that figure should be here as well.